compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEDmaUtils.cpp

```diff
@@ -396,4 +396,37 @@ LogicalResult moveNpuDmaSyncUsersAfterAncestorInSameBlock(
   return success();
 }
 
+LogicalResult moveNpuSourceDmaSyncAfterTargetDmaCpy(RewriterBase &rewriter,
```

Comment:
added some comments in the file, but a bit more explaination here
```ll
%0 = dma_cpy_nd (arg_idx = 0)
dma_wait %0 
%1 = dma_cpy_nd (arg_idx = 1)
dma_wait %1 

%2 = dma_cpy_nd (arg_idx = 0)
dma_wait %2 
%3 = dma_cpy_nd (arg_idx = 1)
dma_wait %3 
```
Somehow during the DMA chaining, I wish to keep `dma_wait %2`. Since everything is unrolled, it is hard to tell if `dma_wait %1 ` and `dma_wait %2 `  have any data dependency. Therefore, the safest option is to stop `%3 = dma_cpy_nd (arg_idx = 1)` being chained with `%1 = dma_cpy_nd (arg_idx = 1)` so that `dma_wait %1` can be kept for synchroinzation. After applying this transformation, we get
```ll
%0 = dma_cpy_nd (arg_idx = 0)
%1 = dma_cpy_nd (arg_idx = 1)
dma_wait %0 
dma_wait %1 

%2 = dma_cpy_nd (arg_idx = 0)
%3 = dma_cpy_nd (arg_idx = 1)
dma_wait %2 
dma_wait %3 
```
where `dma_wait` is grouped together, and it is easy to tell if we wish to  keep `dma_wait %2`, we don't have to keep `dma_wait %1 ` at the same time, so more chaining can happen

Reply:
I don't think you need to be concerned about the `dma_wait` ops while chaining, just about the resources (connections the dma ops are operating on). If two `dma_cpy_nd` are operating on the same connection with no other operation in between, they can be chained and any wait on a `dma_cpy_nd` that is not the last one can be removed.

Reply:
I think the reason is when further lowering the code, all chained DMA operations will share one `push_to_queue`  at the end of chain. This means their `repeat_count`, `direction`, `channel`, etc., must also be identical. Therefore, my current assumption is to build separate chains based on `arg_idx`.

Reply:
`direction` and `channel` are part of the connection and `repeat_count` is part of the `dma_cpy_nd`, so there's no need to depend on `arg_idx` at all. In principal, you could build a chain operating on different args, there's no issue with that.